va-pagination: update for maximum of seven slots

1Copenut opened the pull request:
## Chromatic
<!-- This `tpierce-2199-bug-va-pagination` is a placeholder for a CI job - it will be updated automatically -->
https://tpierce-2199-bug-va-pagination--65a6e2ed2314f7b8f98609d8.chromatic.com

---
## Configuring this pull request
- [x] Link to any related issues in the description so they can be cross-referenced.
- [x] Add the appropriate version patch label (`major`, `minor`, `patch`, or `ignore-for-release`).
    - See [How to choose a version number](https://github.com/department-of-veterans-affairs/component-library#how-to-choose-a-version-number) for guidance.
    - Use `ignore-for-release` if files haven't been changed in a component library package. (ie. only Storybook files)
- [ ] DST Only: Increment the `/packages/core` version number if this will be the last PR merged before a release.
- [ ] Complete all sections below.
- [ ] Delete this section once complete

## Description
Closes https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2199.

This PR adds a `resizeObserver` to the `va-pagination` component to update the number of pagination links on smaller device viewports and zoomed browser windows.

## QA Checklist
- [x] Component maintains 1:1 parity with design mocks
- [x] Text is consistent with what's been provided in the mocks
- [x] Component behaves as expected across breakpoints
- [x] Accessibility expert has signed off on code changes (if applicable. If not applicable provide reason why)
- [x] Designer has signed off on changes (if applicable. If not applicable provide reason why)
- [x] Tab order and focus state work as expected
- [x] Changes have been tested against screen readers (if applicable. If not applicable provide reason why)
- [x] New components are covered by e2e tests; updates to existing components are covered by existing test suite
- [x] Changes have been tested in vets-website using Verdaccio (if applicable. If not applicable provide reason why)

## Screenshots
These are screenshots from a local test on `vets-website` having published the components on Verdaccio (local registry).

**Small screen**
![Screenshot 2025-02-03 124633](https://github.com/user-attachments/assets/76daed46-6173-4793-ab35-c8c23fd82c1b)

---
**Medium screen**
![Screenshot 2025-02-03 124710](https://github.com/user-attachments/assets/43532f40-9704-49bb-a3f1-85dfcae6352b)

---
**Large screen**
![Screenshot 2025-02-03 124737](https://github.com/user-attachments/assets/bbf228d5-027a-436b-9504-67e3277e83f9)

## Acceptance criteria
- [x] QA checklist has been completed
- [x] Screenshots have been attached that cover desktop and mobile screens

## Definition of done
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)

1Copenut commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -338,71 +391,137 @@ export class VaPagination {
-            {renderPages}
-            {endEllipsisAndLastPage}
-            {nextButton}
+            {this.renderPreviousButton(arrowClasses, previousAriaLabel)}
+            {this.renderFirstPage(itemClasses, ellipsisClasses)}
+            {this.renderMiddlePages(ariaLabelSuffix, itemClasses)}
+            {this.renderLastPage(ellipsisClasses, itemClasses)}
+            {this.renderNextPageButton(arrowClasses, nextAriaLabel)}
```
I re-architected these into private functions so we could separate the business logic from the `render()` method. I felt it made the "view layer" easier to read and co-located each piece of the pagination. If there's a better way to render these than individual functions, LMK!

1Copenut commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -125,66 +167,81 @@ export class VaPagination {
+    // The smallest breakpoint can show at most 6 pages, including ellipses
+    if (isMobileViewport) maxPageListLengthState = SHOW_ALL_PAGES;
+
+    // Use cases
```
I lightly refactored the `pageNumbers()` function to use `if, else if, else` logic. I explored using a switch case but didn't feel it was going to scale with multiple props needed to derive the final arrays.

1Copenut commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -233,102 +290,98 @@ export class VaPagination {
+    previousAriaLabel: string,
+  ): JSX.Element {
+    // Skip rendering if the viewport is too narrow to show Previous button
+    if (this.isMobileViewport) return null;
```
I flipped the logic on these new UI render functions. If a condition exists that means this piece of pagination should not be rendered, I returned early using [guard clauses](https://refactoring.com/catalog/replaceNestedConditionalWithGuardClauses.html). This helped me focus the return statements on the exact markup we wanted.

1Copenut reviewed (commented):
Added a few comments to help clarify some of my refactoring decisions.

1Copenut commented:
I tested the change with NVDA, JAWS, and ZoomText on Win11, using Chrome, Firefox, and Edge. Also did basic testing with MS Voice Access to confirm Next, Previous, page numbers clicked as expected.

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.scss:
```diff
@@ -55,3 +56,9 @@ button {
   transition-property: color, background-color;
 }
+
+@container (max-width: 640px) { // Variables won't compile. Matching TSX file for now.
```
> // Variables won't compile. Matching TSX file for now.

What does this mean?

1Copenut replied:
I was having quite a time passing the USWDS `$tablet` variable into this container query. I tried `var(--tablet)` and `$tablet` and `#${tablet}` but they either wouldn't render the CSS rule at all, or failed during compilation. Happy to change it if there's something I'm missing.

jamigibbs replied:
That's really interesting 🤔 I wonder if that's a Stencil limitation? Thanks for explaining! 

Do you mind just mentioning in the comment that the value is for the `--tablet` variable/property?

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -338,71 +391,136 @@ export class VaPagination {
+  /**
+   * =========================================
+   * End pagination render functions
+   * =========================================
```
Nit: I'm personally not a fan of adding these large comment blocks as waypoints for code sections (here and on line 301). Function names should already be identifying what they're for and it seems like they do.

1Copenut replied:
Not a problem, I'll take them out in just a short.

jamigibbs reviewed (commented):
Can you double check the unbounded states behavior? If do the following steps for this [unbounded middle example](https://65a6e2ed2314f7b8f98609d8-ausfvsssad.chromatic.com/?path=/story/uswds-va-pagination--with-unbounded-middle), it seems like I can get to a point where I can't click further than page 5. It doesn't seem like the [current Storybook example](https://design.va.gov/storybook/?path=/story/uswds-va-pagination--with-unbounded-middle) does that.

- Go to [unbounded middle example](https://65a6e2ed2314f7b8f98609d8-ausfvsssad.chromatic.com/?path=/story/uswds-va-pagination--with-unbounded-middle)
- Click 1
- Click 5 and then 6

Can't go any further than 5.

![Screenshot 2025-02-11 at 1 41 55 PM](https://github.com/user-attachments/assets/a6182f81-1ab1-45bf-8d41-3e97245e5598)

1Copenut commented:
@jamigibbs & @Andrew565 I'm finally back with an update to this PR. I added a _significant_ amount of logic to the `pageNumbers()` method to account for more use cases and `maxPagetListLength` prop set to 6, 8, and 10. I also added logic to the resize observer to decide if we're on a small or slightly less small mobile screen. I was having trouble with the middle numbers and two ellipses at small screen size, so it felt like the best way forward.

I've been testing in Storybook for a couple of days and all of the examples do what I expected, and testing a bunch of variants too. Currently testing in `vets-website` locally as much as one can with mocked data.

LMK if the added logic needs split or refactored. Thank you!

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -86,15 +87,73 @@ export class VaPagination {
+   * The number has been reduced from 7 to 6 to show all more consistently
+   * on small screens and zoomed in browser windows.
+   */
+  SHOW_ALL_PAGES: number = 6;
```
Will this comment make sense in the long run? Should this be reverted back to the original comment?

```
If the page total is less than or equal to this limit, show all pages.
```

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -86,15 +87,73 @@ export class VaPagination {
+
+  /**
+   * Mobile viewport width chosen based on USWDS "mobile-lg" spacing unit.
+   * See https://designsystem.digital.gov/design-tokens/spacing-units/
```
I think we would prefer this to reference VADS which also uses the same named/value `mobile-lg` token: https://design.va.gov/foundation/breakpoints

This can be updated in a few different places.

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -86,15 +87,73 @@ export class VaPagination {
 
-  private handlePageSelect = (page, eventID) => {
+  /**
+   * Tabllet viewport width chosen based on USWDS "tablet" spacing unit.
```
Typo: Tabllet

1Copenut commented:
Some screenshots from localhost testing on Your VA Payment History:

**400% zoom at 1280px**

![Screenshot 2025-03-06 092258](https://github.com/user-attachments/assets/3f3f5050-25c4-47ab-9209-03c375bfceec)

---
**300% zoom at 1280px**

![Screenshot 2025-03-06 093329](https://github.com/user-attachments/assets/03db271f-f720-43a5-821f-fc0f08c54f75)

---
**250% zoom at 1280px**

![Screenshot 2025-03-06 093345](https://github.com/user-attachments/assets/296e4677-bf5b-4a4b-a84e-4c7550a9b326)

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -123,68 +182,132 @@ export class VaPagination {
       return makeArray(1, totalPages);
-    }
+    } else if (SHOW_ALL_PAGES < totalPages && totalPages <= maxPageListLength) {
+      console.log('Use case 2');
```
cleanup console.logs in a few different places

jamigibbs reviewed (commented):
@1Copenut Thank you so much for this massive effort! I left a few super minor comments and all of the testing looks pretty good on my end too except one thing I want to check in about.

The [USWDS behavior guidelines](https://designsystem.digital.gov/components/pagination/#behaviors) says:

`The component features a maximum of seven slots.`

But it looks like now there are more than that. Would that be a fairly straight forward to get back in alignment in this PR?

1Copenut commented:
@jamigibbs I appreciate the quick review and comments, thank you. I'll remove the errant console statements and mis-spellings in a quick push and focus on getting the total count back in line with production examples (7 or fewer). Will ping you after I have those updates in place.

1Copenut commented:
Much simpler API now. Removed two props that appeared to be legacy from V1. All pagination layouts now assume there will be a maximum of 7 slots on the screen at any one time. The only hitch is at the smallest breakpoint with 7 pages exactly, the first and last pagination links break out of the container and disappear. I have a `@container` query in the CSS we can use to override behavior in this one case, LMK how DS team would like to proceed!

---
![Screenshot 2025-03-11 110215](https://github.com/user-attachments/assets/16f9d28f-5225-41e7-a69a-7165f4f9f45c)

jamigibbs commented:
Additional context regarding the seven slot max update in this Slack convo: https://dsva.slack.com/archives/C01DBGX4P45/p1741301043783969

danbrady commented:
Hey @1Copenut - The not fitting on mobile is an issue that we have a few issues for already. See [#3003](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/3003) and [#2597](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/2597). Because this issue stems from the USWDS component, we also filed an [issue in their repo](https://github.com/uswds/uswds/issues/5981) too. It looks like they may pick it up soon. It's a deceptively tricky problem considering accessibility guidelines on touch target size, spacing between page links, showing next/previous links, etc in such a small space.

1Copenut commented:
@danbrady This PR should close out your issue [#3003](https://github.com/department-of-veterans-affairs/vets-design-system-documentation/issues/3003) because it removes the `max-page-list-length` prop and solves the next and previous links missing when two ellipses are present.

danbrady commented:
Thanks, @1Copenut! That update is probably what I would have done too since it fixes the layout at 320px for mobile when there's overflow (i.e. ellipses). Just a heads up, if there are exactly 7 pages — so no ellipses — the page buttons will still get cut off at 320px. (Granted, this is likely a rare occurrence.)

@department-of-veterans-affairs/platform-design-system-fe: Just want to note that the USWDS has the issue we submitted for this in the their most recent sprint "2025.3 Feb 17 - Mar 9". So code updates from that may be coming our way.

jamigibbs commented on packages/web-components/src/utils/utils.ts:
```diff
@@ -179,3 +179,21 @@ export function getHeaderLevel(headerInput: number | string): string | null {
+ * See https://dev.to/pubnub/how-fast-is-real-time-human-perception-and-technology-1308
+ * for how the 100ms default was selected.
+ */
+export function debounce(callbackFn: any, delay: number = 100) {
```
If this utility function isn't needed anymore, we can probably remove it.

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -80,21 +75,20 @@ export class VaPagination {
    * `maxPageListLength`
    */
-  @Prop() showLastPage?: boolean = false;
+  // @Prop() showLastPage?: boolean = false;
```
Is this prop also being removed? If so, can you delete instead of commenting it out?

Also make sure to run the `yarn build` step again so that the Stencil generated file gets updated.

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -119,72 +113,67 @@ export class VaPagination {
   private pageNumbers = () => {
     const {
-      maxPageListLength,
+      // maxPageListLength,
```
Cleanup

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.tsx:
```diff
@@ -119,72 +113,67 @@ export class VaPagination {
-    if (totalPages <= this.SHOW_ALL_PAGES) {
+    if (totalPages <= SHOW_ALL_PAGES) {
+      // Use case #1: 7 or fewer total pages
+      console.log('Use case 1');
```
Remove console.logs throughout 🙏🏻

jamigibbs commented on packages/web-components/src/components/va-pagination/va-pagination.scss:
```diff
@@ -55,3 +56,13 @@ button {
+
+@container (max-width: 480px) {
+  // This value matches the USWDS --mobile-lg custom property
+  // See https://designsystem.digital.gov/design-tokens/spacing-units/#spacing-unit-tokens
```
Can we link to the VADS documentation instead in this comment? https://design.va.gov/foundation/breakpoints

jamigibbs commented on packages/web-components/src/components/va-pagination/test/va-pagination.e2e.ts:
```diff
@@ -100,13 +100,13 @@ describe('va-pagination', () => {
   });
 
-  it('renders all page numbers if total pages is less than or equal to 7', async () => {
+  it('renders all page numbers if total pages is less than or equal to 6', async () => {
```
There might be a test for this already but can you double check if there's one that validates it doesn't exceed 7 slots?

jamigibbs commented on packages/web-components/src/components/va-pagination/test/va-pagination.e2e.ts:
```diff
@@ -141,6 +141,93 @@ describe('va-pagination', () => {
+    const page = await newE2EPage();
+    await page.setViewport({
+      // iPhone X
+      width: 375,
```
I like these `setViewport` tests!

danbrady replied:
😮 Wow, this is awesome! Can we add or update mobile to be 320px and still run these tests? That's a iPhone SE (1st gen) resolution and the size we currently design mobile assets for.

jamigibbs reviewed (commented):
This is looking great in my testing @1Copenut! I left just a few comments and it's mostly just cleanup requests. Thanks again for your persistence. 💪

1Copenut commented:
@jamigibbs Ready for another review. I updated and cleaned up files as requested. Added a couple of tests for first and last page logic testing too.
